Davids Theme (Update to OpenResume UI Theme)

DavidSchurer opened the pull request:
This is my own theme that I have worked on, there is still more to add, but what I have done is essentially added navigation tabs for each resume section such as work experience, skills, education, etc. I implemented it in a carousel fashion so the user can click on either of the two arrows to cycle through the different resume sections. I also added an input filter for sorting skills by either years of experience or specific skills.

missionmike commented on src/types/index.ts:
```diff
@@ -243,3 +243,18 @@ export interface Education {
 }
+
+// Certification type for the Certifications section
+export interface Certification {
```
This is another important concept that the backend will need to support before we can include it in themes.

If you'd like, you can move the `ProjectSection` interface into this `index.ts` file, and relabel it as `FeaturedProject`. Then, I can base the backend work off these new interfaces and let you know when they're supported so their UI can be reintroduced to the theme.

missionmike commented on tsconfig.json:
```diff
@@ -12,18 +16,28 @@
     "resolveJsonModule": true,
     "isolatedModules": true,
-    "jsx": "react-jsx",
+    "jsx": "preserve",
```
Was this changed for a reason? I'm not sure if it has any effect, just curious.

missionmike replied:
> I did end up changing this as I have heard that next.js recommends letting its own compiler handle handle JSX transformation rather than TypeScript, if you want me to I can always revert it back to react-jsx just let me know!

Thanks! I updated it to use `preserve` in the `develop` branch, going to see if it works the same.

In this repo, the Next.js portion is mainly used to help us build themes and share example links to demo them. 

However, during the [build step when this repo is deployed](https://github.com/missionmike/openresume-theme/blob/develop/.github/workflows/npm-publish.yml), it bundles it differently using [tsup](https://github.com/missionmike/openresume-theme/blob/develop/package.json#L21) (see [tsup on npm](https://www.npmjs.com/package/tsup)) instead of `next build`... this uses the `tsconfig.build.json` file, which extends `tsconfig.json`. The bundle is then published to npm and served as `@openresume/theme` for installation/reuse by the backend.

I don't even remember why I ended up with `jsx-react` instead of `preserve` in the tsconfig... we'll see if `preserve` still works when we publish!

missionmike commented on src/theme/index.ts:
```diff
@@ -1,2 +1,4 @@
 export { ThemeDefault } from "./default/ThemeDefault";
 export { ThemeDefaultPDF } from "./default/ThemeDefaultPDF";
+export { ThemeDavidsTheme } from "./Davids-Theme/DavidsTheme";
```
How do you feel about calling it `ThemeDavids`?

I'm thinking the many instances of "theme" become redundant.

Then the folder name could simply be `./Davids/`.

I'm find with calling the theme "David's Theme" though, of course. I just think it becomes redundant in the code if we include the word "theme."

I also anticipate many. more themes being added, which could follow the same naming convention (e.g., including "Theme" in the printed title is fine, but it doesn't need to be in the code).

Thoughts?

DavidSchurer replied:
Sounds good, ill just rename it to Davids to avoid confusion with other redundant themes with the word "theme" in them.

DavidSchurer replied:
I think ill just leave it as ThemeDavids for now but then later on I will change it to a different name, such as Davids

missionmike replied:
@DavidSchurer lol, naming things is hard. Especially otherwise-simple things like a theme... it's early in the project, and I didn't really document what I had in mind for the naming convention for code and UI... that's my bad!

Here's what I'm thinking:

The theme "slug" doesn't need the word "theme" in it, since we know it's a theme slug and it's a reference to a theme. (I should probably rename the type `ThemeName` to `ThemeSlug` for clarity... perhaps I'll do this after we merge your PR).

The slug is what appears in the folder structure, and is what gets used to determine which theme to serve, e.g.:

- `src/theme/default` --> `"default"` is `ThemeName`
- `src/theme/davids' --> `"davids"` is `ThemeName`

Then, within the folder, the primary point for the theme components are:

- `src/theme/default/DefaultTheme`
- `src/theme/default/DavidsTheme`

The reason these components have `Theme` in them is to provide clarity on what they are when they're imported elsewhere. If we're importing `Default` as a component, it's not really descriptive as to what it is.... but importing `DefaultTheme` tells the developer on the other end what they're importing, specifically.

Lastly, in the UI and nav menu, these themes will be presented as:

- "Default Theme"
- "David's Theme"

Let me know if this sounds good on your end, or if you have questions/suggestions. Thanks!

missionmike commented on src/theme/Davids-Theme/sampleData.json:
Did you include the new custom data here because of the addition of the new `projects` and `certifications` data?

I'm thinking you could stash this until we can add the support for those, and use the default sample data (feel free to modify the sample data if you'd like, it can be shared between the default theme and yours).

I'm hoping we can share the same sample data across multiple themes. I'm envisioning a feature where users can be viewing a resume on the demo site, then toggle the chosen theme from a dropdown to rearrange the same data into a different theme.

DavidSchurer replied:
Yes, that is correct, I did include the addition of projects and certifications, ill go ahead and stash those until there is support for them!

missionmike commented on src/theme/Davids-Theme/DavidsTheme.tsx:
```diff
@@ -0,0 +1,569 @@
+  }
+  
+  export const ThemeDavidsTheme = ({
+    user = {
```
The data for the theme gets passed in, so there's no need to set the default data here.

DavidSchurer replied:
thanks for the catch, i fixed this!

missionmike commented on src/theme/Davids-Theme/DavidsTheme.tsx:
```diff
@@ -0,0 +1,569 @@
+    certifications?: Certification[];
+  }) => {
+    /* ----------  FALLBACK SAMPLE DATA ---------- */
+    const defaultProjects: ProjectSection[] = [
```
We don't need to set default data for the themes within the theme itself. We should treat these as optional sections - if a user doesn't populate data for a section within the theme, then that section shouldn't render at all.

missionmike commented on src/theme/Davids-Theme/DavidsTheme.tsx:
```diff
@@ -0,0 +1,569 @@
+
+    /* ----------  SECTION DATA ---------- */
+    // Add state for skill filter and filter type
+    const [skillFilter, setSkillFilter] = useState("");
```
This is an interesting addition! Do you think we should only show the filter if there's more than like 20 skills listed? I imagine users won't need to use this if there's 5-10ish skills on the page... but I can see it being useful if there's a whole bunch to read through.

I like the concept, definitely thinking outside the box here on the UX!

DavidSchurer replied:
I feel this might be useful for employers to see how many years of experience someone has for a specific skill, I was trying to think of any UI additions to make but I might keep this implementation in here but just refine it.

missionmike commented on src/theme/Davids-Theme/DavidsTheme.tsx:
```diff
@@ -0,0 +1,569 @@
+    }, [skillFilter, skillsForUser, filterType]);
+    
+
+    const sections: { label: string; render: JSX.Element | null }[] = [
```
This is an interesting setup; the array of sections with label and accompanying component tree... do you think each section could be extracted into its own component? Example:

```
{
  label: "Skills",
  render: <Skills skillsForUser={skillsForUser} />, // You may be able to move the filter component directly in here, so you don't need to pass in the state/setState.
},
// ...
```

Then move the inner components into `./components/Skills.tsx`

Also, I like how you re-used the default theme `Skills` component to render the results!

There's plans to incorporate more unit testing with Jest, so having smaller, separate components will make it easier to test.

DavidSchurer replied:
Yes absolutely! Let me restructure this so each section has its own component to make the codebase more maintainable and easier to test.

missionmike commented on src/theme/Davids-Theme/DavidsTheme.tsx:
```diff
@@ -0,0 +1,569 @@
+      {
+        label: "Projects",
+        render: projectData.length ? (
+          <Box>
```
When extracting the components, feel free to keep this Projects component available, but we don't need to render it in the theme yet. It can live in `./components/FeaturedProjects.tsx` in prep for when we have the featured projects backend up & running.

DavidSchurer replied:
with my new pr request, ill make sure to not render projects and certifications but keep them as their own components and not render them into the theme

missionmike commented on src/theme/Davids-Theme/DavidsTheme.tsx:
```diff
@@ -0,0 +1,569 @@
+    }
+  
+    /* ----------  DARK-NAVY THEME ---------- */
+    const navyTheme = createTheme({
```
Nice usage of the MUI theme functionality!

Please give it a pass in Dark mode as well. Some of the colors made text hard to read in dark mode:

<img width="554" alt="image" src="https://github.com/user-attachments/assets/62161f29-fd06-4528-bdca-69f372337e93" />

missionmike commented on package-lock.json:
I'm not sure why this change happens... when you get a chance, can you please try to update your `develop` branch from the origin's `develop`? I think via the "Sync Fork" button:

<img width="186" alt="image" src="https://github.com/user-attachments/assets/835cbaa8-eb17-4fad-80d6-a40b73b0d251" />

missionmike reviewed (changes requested):
@DavidSchurer this is such an excellent start! I'm super excited for this progress, and I'm grateful that you've worked on this new theme. I love the carousel concept and the new ideas re: Projects and Certifications. I'll get those prioritized for the backend so we can hopefully implement those sooner than later... please see comments and suggestions.

Looking forward to the next iteration!

If you have any questions at all, please reach out here. Thank you!

aweb3dad commented:
Oh man, it's been so long since I've looked at a PR, and I'm realizing why I don't like PRs. It takes forever to build an application, and the concern of adding bad code to the project is always a big concern.

I'd like to aim for us to install some sort of summarizer here. Gonna ask Reddit if it's possible, that way it's not so difficult to read a PR and keep up from this level, that way I can also provide some guidance here.

Development isn't able to be as slow as it used to be, especially with ai trying to take over the review process as well... so I'm just watching and seeing where we can optimize, since I wanna bring some new skillsets to organizations when I start applying for jobs more. And we might be able to sell these interviewers on the fact that they're code review process sucks... and how we code review is much more efficient... so finding the method right now

missionmike commented:
> Oh man, it's been so long since I've looked at a PR, and I'm realizing why I don't like PRs. It takes forever to build an application, and the concern of adding bad code to the project is always a big concern.
> 
> I'd like to aim for us to install some sort of summarizer here. Gonna ask Reddit if it's possible, that way it's not so difficult to read a PR and keep up from this level, that way I can also provide some guidance here.
> 
> Development isn't able to be as slow as it used to be, especially with ai trying to take over the review process as well... so I'm just watching and seeing where we can optimize, since I wanna bring some new skillsets to organizations when I start applying for jobs more. And we might be able to sell these interviewers on the fact that they're code review process sucks... and how we code review is much more efficient... so finding the method right now

GitHub Copilot has a summarizing feature, but I just cancelled that subscription because I wasn't using it after moving to Cursor. I think 90% of code review time could be saved if initial PRs always consider a few things:
- Are variables always named consistently, clearly, indicating what they're for similarly across components?
- Are components small, testable, well-named?
- Can anything be simplified or refactored?
- Are comments and code styles consistent throughout?
- Were tests added to cover anything new? (I'm still working on adding these for the default theme in this case, so I haven't set a proper example yet)
- Did CI pass?

Some of these principles only come about through practice and after gaining familiarity with a codebase. Perhaps an auto-reviewer could scan PRs and ask itself these questions, then provide immediate recommendations to the author for an initial iteration.

There's always going to be a human in the loop though, especially on a codebase where I envision 100s of themes being available in the future. Quality typically takes time.

DavidSchurer commented:
With this latest PR commit, I implemented some new functionality to the UI, I added a QR code generator, which I believe would make it easier for users to share their resume around on other platforms, especially with a download button where they can download it as a PNG file and use it on other sites to share and advertise their online resume/portfolio.
<img width="808" alt="qrcode_darkmode" src="https://github.com/user-attachments/assets/61c8f075-b4d7-4393-af6e-24ac524c03ee" />
<img width="828" alt="qrcode_lightmode" src="https://github.com/user-attachments/assets/05918afd-3b9d-4abb-b845-6f32520511c0" />

missionmike commented on src/theme/themeNavItems.tsx:
```diff
@@ -12,4 +12,9 @@ export const themeNavItems: NavItem[] = [
   },
+  {
+    text: "David's Theme",
+    icon: <FeedIcon />,
```
Feel free to choose any other available icon as well. It can be something unique, to help differentiate your theme from others.

missionmike commented on src/types/index.ts:
```diff
@@ -18,7 +18,21 @@ export type ThemeAppearance = "dark" | "light";
  * on the OpenResume site, and should be unique to each theme.
  */
-export type ThemeName = "default"; // Add more themes here, e.g. "default" | "my-theme" | "another-theme"
+export type ThemeName = "default" | "davids-theme"; // Added new Davids theme
```
Per: https://github.com/missionmike/openresume-theme/pull/11/files#r2108211799

```suggestion
export type ThemeName = "default" | "davids"; // // Add more themes here, e.g. "default" | "my-theme" | "another-theme"
```

DavidSchurer commented:
> > With this latest PR commit, I implemented some new functionality to the UI, I added a QR code generator, which I believe would make it easier for users to share their resume around on other platforms, especially with a download button where they can download it as a PNG file and use it on other sites to share and advertise their online resume/portfolio.
> 
> This is an awesome idea!
> 
> I'm curious about the intended UX.
> 
> When the QR code is displayed on the resume, is it intended to be an easier way for someone to scan and view the same page on their phone, so that they could then share the URL from their phone via copy/paste? Are there other use cases for it?
> 
> Once use cases are fleshed out, I'd like to consider eventually moving this into a globally-shared component that any theme could opt-in to use.
> 
> Similar to how your theme uses the `SkillItem` and other things from the default theme - I might extract those to a "common theme components" folder that shares with any theme. A QR code generator could be a nice addition to the common components.

Yes, the main use case is for someone to be able to scan it to view the resume on their mobile device to allow for easy sharing/access, it could also be downloaded and exported as a png image so users could display it on a third party website which would directly link the user to their resume page, lets say if they had thrown this qr code on their github page or linkedin page so people could scan the QR code to easily pull up their resume.

aweb3dad commented:
Anything we can do to get this merged in? I don't want to sit too long on a PR, especially a 2 year old one @missionmike '😂 referencing a post you made on linkedin". My goal here is to see how we can optimize development to bolster the look of our profile pages while also having a live demo of what we're building, so if I have to pay for like an elastic beanstalk to host this PR in an isolated environment for us to test let me know.

I don't mind doing some configuration to get it so that folks can see their preview and experience it before merging it in. I know it may sounds crazy for an open-source project, but hell now is the time to do corporate level stuff inside of an open-source project, especially since I can find the folks to help us. Especially since we can learn from each other simultaneously.
